### ISSUE TITLE
ANSIBLE_BECOME_EXE not working with Ansible 2.4

### ISSUE THREAD

**Issue by bernardoVale** [ansible/ansible#32157]
`ANSIBLE_BECOME_EXE` not honored in Ansible 2.4.

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
privillege escalation

##### ANSIBLE VERSION
```
ansible 2.4.0.0
  config file = /Users/bvale/wsgc/jenkins2-install/ansible.cfg
  configured module search path = [u'/Users/bvale/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.13 (default, Dec 18 2016, 07:03:39) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

##### OS / ENVIRONMENT
N/A

##### SUMMARY
Looks like ansible is ignoring `ANSIBLE_BECOME_EXE` environment variable in Ansible 2.4 or the behavior of this variable had changed.

**Ansible 2.3**

```
ansible 2.3.0.0
  config file = /foo/bar/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Dec 18 2016, 07:03:39) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
````

ID command with `become_exe` snippet:

```
<myserver> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=a_bvale -o ConnectTimeout=10 -o ControlPath=/Users/bvale/.ansible/cp/ae87db0414 -tt jnkrck-vdcn001 '/bin/sh -c '"'"'sudo su -  desired_user -c '"'"'"'"'"'"'"'"'/bin/sh -c '"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-ouqocexdwfwclipzardqbyzkgpcltuzp; /usr/bin/python /tmp/ansible-tmp-1508954464.54-50086839833054/command.py'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"''"'"'"'"'"'"'"'"' && sleep 0'"'"''
```

**Ansible 2.4**

ID command with `become_exe` snippet:

```
<myserver> SSH: EXEC ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=myuser -o ConnectTimeout=10 -o ControlPath=/Users/bvale/.ansible/cp/ae87db0414 -tt myserver '/bin/sh -c '"'"'su  desired_user -c '"'"'"'"'"'"'"'"'/bin/sh -c '"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-vfjalzrurfjclyabvzfzrdgpohoinirh; /usr/bin/python /tmp/ansible-tmp-1508954441.8-223685343608311/command.py'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"'"''"'"'"'"'"'"'"'"' && sleep 0'"'"''
```

Ansible 2.4: `su  desired_user -c `
Ansible 2.3: `sudo su -  desired_user -c `

##### STEPS TO REPRODUCE
Run a simple ad-hoc shell command becoming a specific user using `ANSIBLE_BECOME_EXE="sudo su-"`.

```
ANSIBLE_BECOME_EXE="sudo su -" ansible -m shell -a 'id' --become --become-method su --become-user desired_user myserver
```

##### EXPECTED RESULTS
Should return user id output:

Ansible 2.3 result:

```
uid=169(desired_user) gid=169(desired_user) groups=169(desired_user) context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
```

##### ACTUAL RESULTS

Ansible 2.4 result:

```
Timeout (12s) waiting for privilege escalation prompt:
```

**Comment by bernardoVale** [comment]
Quick update. Tested with current code on develop and its working fine

**Comment by jborean93** [comment]
If it works fine with the current code then it has probably been fixed and backported to 2.4. You can test out the rc2 for 2.4.1 to see if it will be available in that release https://groups.google.com/forum/#!topic/ansible-project/Rttds8wUJLM. 2.4.1 should be out sometime soon and if it does work please close the issue.

**Comment by bernardoVale** [comment]
Yep. I've also tested with the last tag. I think we should wait for the official release 2.4.1 before closing this issue.

**Comment by jborean93** [comment]
You can test out rc2 which is going to be the official release unless any major blockers are found (which is unlikely)

**Comment by madeinoz67** [comment]
have just updated to 2.4.1 and still experiencing this issue

```
ansible 2.4.1.0
  config file = /Users/seaton/Documents/ansible-production/ansible.cfg
  configured module search path = [u'/Users/seaton/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/ansible
  executable location = /Library/Frameworks/Python.framework/Versions/2.7/bin/ansible
  python version = 2.7.9 (v2.7.9:648dcafa7e5f, Dec 10 2014, 10:10:46) [GCC 4.2.1 (Apple Inc. build 5666) (dot 3)]
```

**Comment by madeinoz67** [comment]
It also seems to be isolated to MacOS environment see https://github.com/StackStorm/ansible-st2/issues/180

I'm seeing that ansible 2.4.0.0 works on Centos7  same with 2.4.1.0, neither on MacOS.

**Comment by bernardoVale** [comment]
I can confirm that this is no longer an issue in 2.4.1. I will close this issue!

**Comment by es1o** [comment]
I still have this issue in 2.4.3.0 - please reopen this bug.

Both become_exe in ansible.cfg and ANSIBLE_BECOME_EXE doesn't work.
`module_stdout": "\r\nSorry, user my_useris not allowed to execute '/bin/su - user-c /bin/sh -c 'echo BECOME-SUCCESS-tfczlpeduednzqbuuumjnwytdlgkghxg; /usr/bin/python /tmp/ansible-tmp-1519391857.7-47182692798888/command.py'' as root on host.\r\n",`

Why ansible trying use /bin/su instead of sudo su -?

**Comment by jmcgrath207** [comment]
I also have a issue where the become_exe does not work in 2.4.3.0

```
- name: "Install Packages | Install  {{  System_Packages | join(', ') }}  Packages"
  yum: state=present  name={{ item }}
  with_items:
    - '{{ System_Packages }}'
  become: true
  become_exe: /usr/bin/ss
  ignore_errors: yes
```

**Comment by ravibhure** [comment]
I am also facing issue in 2.4.3.0, my user having /etc/sudoers  entry to use  `/bin/su -`  without password

```
$ ansible -i /tmp/tamtam www-dev -u ravi.bhure -m shell -a "tail /var/log/yum.log" --become
www-dev | FAILED! => {
    "changed": false, 
    "module_stderr": "Shared connection to www-dev closed.\r\n", 
    "module_stdout": "sudo: a password is required\r\n", 
    "msg": "MODULE FAILURE", 
    "rc": 1
}
```

```
$ ansible -i /tmp/tamtam tam-dev-be1 -u ravi.bhure -m shell -a "tail /var/log/yum.log" --become-method=su  --become-user=root --become
www-dev | FAILED | rc=-1 >>
Timeout (12s) waiting for privilege escalation prompt: 
```

**Comment by sivel** [comment]
Issues with `ANSIBLE_BECOME_EXE` were indeed resolved in 2.4.1 via 71c646fe593d4658df4a9181aaf7cb56b99cd091

If you are still experiencing other issues on recent versions of ansible (v2.5 or newer) please feel free to open a new issue.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
